<a href="https://colab.research.google.com/github/disiree31-abel/cs82a-portfolio/blob/main/module3_lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd

df = pd.read_csv("usarrests.csv")

print("Shape:", df.shape)
display(df.head())
print("\nData types:")
print(df.dtypes)

Shape: (50, 4)


,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91



Data types:
Unnamed: 0     object
Murder        float64
Assault       float64
UrbanPop        int64
dtype: object


In [4]:
df.describe()

,Murder,Assault,UrbanPop
count,50.00000,49.000000,50.00000
mean,7.78800,182.183673,74.20000
std,4.35551,130.877435,73.40828
min,0.80000,45.000000,6.00000
25%,4.07500,109.000000,53.25000
50%,7.25000,159.000000,66.00000
75%,11.25000,249.000000,77.75000
max,17.40000,879.000000,570.00000


In [5]:
summary = df.describe().T
display(summary)

,count,mean,std,min,25%,50%,75%,max
Murder,50.0,7.788000,4.355510,0.8,4.075,7.25,11.25,17.4
Assault,49.0,182.183673,130.877435,45.0,109.000,159.00,249.00,879.0
UrbanPop,50.0,74.200000,73.408280,6.0,53.250,66.00,77.75,570.0


In [6]:
print("Maximum values:")
print(df.max(numeric_only=True))

Maximum values:
Murder       17.4
Assault     879.0
UrbanPop    570.0
dtype: float64


In [7]:
print("UrbanPop maximum:", df["UrbanPop"].max())

UrbanPop maximum: 570


In [11]:
display(df[df["UrbanPop"] > 100])

,Unnamed: 0,Murder,Assault,UrbanPop
14,Iowa,2.2,56.0,570


In [12]:
print("Murder minimum:", df["Murder"].min())
print("Murder maximum:", df["Murder"].max())

print("Assault minimum:", df["Assault"].min())
print("Assault maximum:", df["Assault"].max())

print("UrbanPop minimum:", df["UrbanPop"].min())
print("UrbanPop maximum:", df["UrbanPop"].max())

Murder minimum: 0.8
Murder maximum: 17.4
Assault minimum: 45.0
Assault maximum: 879.0
UrbanPop minimum: 6
UrbanPop maximum: 570


In [13]:
print("Negative Murder values:")
display(df[df["Murder"] < 0])

print("Negative Assault values:")
display(df[df["Assault"] < 0])

print("UrbanPop over 100:")
display(df[df["UrbanPop"] > 100])

Negative Murder values:


,Unnamed: 0,Murder,Assault,UrbanPop


Negative Assault values:


,Unnamed: 0,Murder,Assault,UrbanPop


UrbanPop over 100:


,Unnamed: 0,Murder,Assault,UrbanPop
14,Iowa,2.2,56.0,570


Flagged Values
1. Iowa — UrbanPop = 570

I found a problem with Iowa's UrbanPop value because it is listed as 570. Since UrbanPop is a percentage, it should not be higher than 100, so I consider this value impossible.

I would flag this value and check the original data to see if it was entered incorrectly. I would not change it myself without knowing what the correct value is.

- Extreme Values

Some of the crime numbers are higher than others, but that does not automatically mean they are wrong. A high crime rate can still be a real value, so I would consider it extreme instead of impossible unless there is evidence that it is an error.

For the Iowa UrbanPop value, though, the number is different because 570% is not possible. I would leave the other extreme values alone unless I found more information showing that they were incorrect.